opal/rocm/common_rocm.c

```diff
@@ -0,0 +1,319 @@
+    if (hipSuccess == err) {
+        if (hipMemoryTypeDevice == srcAttr.memoryType || hipMemoryTypeUnified == srcAttr.memoryType) {
+            srcFlag = true;
+        }
+    }
+    err = hip_funcs.hipPointerGetAttributes(&dstAttr, dst);
+    if (hipSuccess == err) {
+        if (hipMemoryTypeDevice == dstAttr.memoryType || hipMemoryTypeUnified == dstAttr.memoryType ) {
+            dstFlag = true;
+        }
+    }
+    if (srcFlag || dstFlag) {
+        return true;
+    }
+    
+    return false;
+}
+
+int mca_common_rocm_memcpy_sync(void *dst, void *src, size_t nBytes)
+{
+
+    if (!rocm_initialized) {
+        mca_common_rocm_support_init();
+    }
+
+    if (nBytes < mca_common_rocm_memcpy_limit) {
+        memcpy(dst, src, nBytes);
+        return OPAL_SUCCESS;
+    }
+
+    if (mca_common_rocm_memcpy_async) {
+        HIP_CHECK(hip_funcs.hipMemcpyAsync (dst, src, nBytes, hipMemcpyDefault,
+                                            common_rocm_MemcpyStream));
+        HIP_CHECK(hip_funcs.hipStreamSynchronize(common_rocm_MemcpyStream));
+    }
+    else {
+        HIP_CHECK(hip_funcs.hipMemcpy (dst, src, nBytes, hipMemcpyDefault));
+    }
+
+    return OPAL_SUCCESS;
+}
+
+void *mca_common_rocm_memcpy(void *dst, const void *src, size_t nBytes, opal_convertor_t *pConvertor)
+{
+
+    if (!(pConvertor->flags & CONVERTOR_ROCM) || nBytes < mca_common_rocm_memcpy_limit) {
+        //For short messages a regular memcpy is faster than hipMemcpy or hipMemcpyAsync
+        return memcpy(dst, src, nBytes);
+    }
+
+    if (mca_common_rocm_memcpy_async) {
+        HIP_CHECK_RETNULL(hip_funcs.hipMemcpyAsync (dst, src, nBytes, hipMemcpyDefault, common_rocm_MemcpyStream));
+        HIP_CHECK_RETNULL(hip_funcs.hipStreamSynchronize(common_rocm_MemcpyStream));
+    }
+    else {
+        HIP_CHECK_RETNULL(hip_funcs.hipMemcpy (dst, src, nBytes, hipMemcpyDefault));
+    }
+
+  return dst;
+}
+
+int mca_common_rocm_memmove(void *dst, void *src, size_t nBytes)
+{
+    if (!rocm_initialized) {
+        mca_common_rocm_support_init();
+    }
+
+    char *tmp=NULL;
+    HIP_CHECK(hip_funcs.hipMalloc((void **)&tmp, nBytes));
+    if (mca_common_rocm_memcpy_async) {
+        HIP_CHECK(hip_funcs.hipMemcpyAsync (tmp, src, nBytes, hipMemcpyDefault, common_rocm_MemcpyStream));
+        HIP_CHECK(hip_funcs.hipMemcpyAsync (dst, tmp, nBytes, hipMemcpyDefault, common_rocm_MemcpyStream));
+        HIP_CHECK(hip_funcs.hipStreamSynchronize (common_rocm_MemcpyStream));
+    } else {
+        HIP_CHECK(hip_funcs.hipMemcpy (tmp, src, nBytes, hipMemcpyDefault));
+        HIP_CHECK(hip_funcs.hipMemcpy (dst, tmp, nBytes, hipMemcpyDefault));
+    }
+
+    HIP_CHECK(hip_funcs.hipFree(tmp));
+    return OPAL_SUCCESS;
+}
+
+void mca_common_rocm_register_mca_variables(void)
+{
+
+    /* Set verbosity in the crocm related code. */
+    mca_common_rocm_verbose = 0;
+    (void) mca_base_var_register("ompi", "mpi", "common_rocm", "verbose",
```

Comment:
it doesn't appear you use this variable?

Reply:
I can remove it if you want, but I am pretty confident that I will need it in future iterations of this code, its really just the starting point.